fix(core): certain template strings could not be resolved in configs

edvald opened the pull request:
This addresses a design issue, where runtime values were being computed
to interpolate module configurations. Because service names and outputs
cannot really be resolved until after all module configs are resolved,
we needed to remove the `services` template string context, as well
as fix the module resolution flow.

BREAKING CHANGE:

Module configurations using the `services` template key need to be
updated to use `modules` instead.

The (admittedly poorly supported) google-cloud-function module type has
been changed to include only one function per module. This is more
consistent with other module types, and avoids complex refactoring
to fit with the changes in the templating context.

thsig commented on garden-service/src/plugins/google/google-app-engine.ts:
```diff
@@ -20,34 +17,42 @@ import {
+
+        // TODO: we may want to pull this from the service status instead, along with other outputs
+        const project = params.ctx.provider.config.project
+        const endpoint = `https://${GOOGLE_CLOUD_DEFAULT_REGION}-${project}.cloudfunctions.net/${config.name}`
```
Should the region here be override-able via config?

edvald commented:
Added another commit, which I think takes this thing home:

**fix(core): using module version in templates didn't work with watch**

This is a pretty major refactor, where resolving raw configs is
deferred to usage time, and the handling of resolved configuration is
moved to the ConfigGraph class (née DependencyGraph). Basically the
Garden class now holds the raw configuration, and it is fully resolved
when building a new instance of ConfigGraph, which should then be done
whenever module sources are modified (otherwise resolved template
strings, e.g. version tags, may be stale).

Also moved some handling of actions to the ActionHandler to further
reduce the size of the Garden class.

thsig commented on garden-service/src/garden.ts:
```diff
@@ -146,16 +108,13 @@ const scanLock = new AsyncLock()
 
   private readonly loadedPlugins: { [key: string]: GardenPlugin }
   private moduleConfigs: ModuleConfigMap
+  private pluginModules: ModuleConfig[]
```
Suggest renaming this to `pluginModuleConfigs`.

edvald commented on garden-service/src/garden.ts:
```diff
@@ -516,123 +472,55 @@ export class Garden {
-    ]
-
-    return sortBy(uniqByName(modules), "name")
+  async getConfigGraph() {
```
I added a note in the function doc to explain, but maybe `resolveConfigGraph()` would make sense?
